src/api/app/controllers/trigger_controller.rb

```diff
@@ -1,16 +1,71 @@
 class TriggerController < ApplicationController
+  validate_action rebuild: { method: :post, response: :status }
+  validate_action release: { method: :post, response: :status }
   validate_action runservice: { method: :post, response: :status }
 
+  before_action :require_valid_token
+
   #
   # This controller is checking permission always only on the base of tokens
   #
   skip_before_action :extract_user
   skip_before_action :require_login
 
-  # github.com sends a hash payload
-  skip_before_action :validate_params, only: [:runservice]
+  def rebuild
+    Backend::Api::Sources::Package.rebuild(@pkg.project.name, @pkg.name)
+    render_ok
+  end
+
+  # rubocop:disable Lint/NonLocalExitFromIterator
+  def release
+    if params[:project].present?
+      render_error message: 'Token must define the release package',
+                   status: 403,
+                   errorcode: 'no_permission'
+      return
+    end
+
+    matched_repo = false
+    User.current = @token.user
+    @pkg.project.repositories.each do |repo|
+      repo.release_targets.each do |releasetarget|
+        next unless releasetarget.trigger == 'manual'
+        unless @token.user.can_modify?(releasetarget.repository.project)
```

Comment:
`repo.project`

You are using the project several time (some of them as `@pkg.project`). I think it is a good idea to assign the project to a variable before the loop :wink:

Reply:
because it's a nitpick, I would let as it is, since the number of matches is really specific, I would keep it as @adrianschroeter did it and optimize it if necessary later... what do you think?